### PR TITLE
feature/s21_strtok

### LINKED ISSUE #32: `s21_strtok` - Tokenizes str using delim as separators.
## Description:  
The task is to implement the `s21_strtok` function, which breaks the string `str` into a series of tokens separated by `delim`.  

## Requirements:  

- [x] The function must be named `s21_strtok` and conform to the C11 standard.  
- [x] The implementation must be placed in the `s21_string.h` library.  
- [x] Follow the Google C++ Style Guide for code formatting.  
- [x] Ensure the function works with z-strings made of single-byte characters in ASCII encoding.  
- [x] Write unit tests using the Check library to verify the correctness of the implementation. The tests must cover at least 80% of the function (checked using gcov).  
- [x] The function must handle edge cases such as consecutive delimiters, leading/trailing delimiters, and `NULL` input for subsequent calls.  

## Expected Behavior:  
- For input `"Hello,world,!"` with delimiter `","`, the function should return tokens `"Hello"`, `"world"`, and `"!"` in successive calls.  
- For input `"   Hello   world   "` with delimiter `" "`, the function should return tokens `"Hello"` and `"world"` (ignoring leading/trailing spaces).  
- For input `""` (empty string), the function should return `NULL`.  

## Additional Notes:  
- The function must maintain state between calls (like the standard `strtok`).  
- Include tests for cases where the input string contains only delimiters or no delimiters.  
- Ensure the function does not modify the original string except for replacing delimiters with null terminators.

### PR DESCRIPTION
**Pull request by DarkDomian**
## Closed #32 | **Overview**  
This PR introduces the `s21_strtok()` function, a custom implementation of the standard `strtok()` function for the **string+** library. The function provides tokenization of strings using specified delimiters while maintaining compatibility with the original behavior.

## **Key Features**  
- [X] **Full compliance** with standard `strtok()` behavior:  
   - Supports sequential calls with `S21_NULL`  
   - Handles multiple delimiters  
   - Properly manages internal state between calls  

- [X] **Comprehensive test coverage** including:  
   - Basic tokenization (space delimiter)  
   - Complex delimiters (punctuation, mixed chars)  
   - Edge cases (empty string, no delimiters found)  

- [X] **Memory-safe** implementation with proper string termination    

### **Implementation Highlights**  
- Uses `s21_strspn()` for delimiter skipping  
- Leverages `strpbrk()` and `strchr()` (to be replaced with `s21_` versions later)  
- Maintains internal state via `static` variable `olds`  

### **Dependencies**  
- Requires `s21_strspn()` helper function (included)  
- Will eventually use `s21_strpbrk()` and `s21_strchr()` (marked as TODOs)  

### **Test Cases**  
1. **Basic Tokenization** (`test_strtok_space_delim`)  
2. **Complex Delimiters** (`test_strtok_comma_delim`)  
3. **Mixed Delimiters + NULL Calls** (`test_strtok_mixed_delims_with_null`)  
4. **Edge Cases** (`test_strtok_empty_string`, `test_strtok_no_delims`)  

### **How to Verify**  
```bash
make strtok.test
```
## Short describe of function's workflow:
signature: `char *s21_strtok(char *str, const char *delim)`
Parse `str` into tokens separated by characters in `delim`. If `str` is `S21_NULL`, the last string `strtok()` was called with is used. For example:
```cpp
    char s[] = "-abc-=-def";
    x = strtok(s, "-");		// x = "abc"
    x = strtok(NULL, "-=");		// x = "def"
    x = strtok(NULL, "=");		// x = NULL
            // s = "abc\0=-def\0"

```

Pay attention: that function rewrite existing string (that's why every test check memory too)

### CODE REVIEW

**Comment by ParthenGithub** [comment]
Why it's draft?

File: src/s21_strtok.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    if (*a == '\0')
+      return count;
+    else
+      ++count;
+  }
+
+  return count;
+}
+
+// TODO: add `s21_` prefix to `strpbrk` and `strchr` when they will be done
```
Comment (ParthenGithub) [inline src/s21_strtok.c:34]: Mayhaps you can do 
`#define s21_strpbrk strpbrk`
In file's beginning, so later it will be easier to replace

File: tests/test_strtok.c [hunk 2]
```diff
@@ -0,0 +1,129 @@
+#include <stdlib.h>
+#include <string.h>
+
+#include "../headers/s21_string.h"
+#include "./suites.h"
+
+#define ALICE_TEXT                                                                 \
```
Comment (ParthenGithub) [inline tests/test_strtok.c]: Define? Seriously?)

**Review by ParthenGithub — APPROVED** [review]

File: src/s21_strtok.c (hunk 1 above)
Comment (ParthenGithub) [reply src/s21_strtok.c:34]: Forget it. I tried to make my it myself and it's so much harder and messy

**Review by gashturik — APPROVED** [review]

**Comment by gashturik** [comment]
The work is done!!!
